libraries/reconstruction/angular_sph_alignment.cpp

```diff
@@ -0,0 +1,597 @@
+{
+    // Read the reference volume
+    V.read(fnVolR);
+    V().setXmippOrigin();
+    Xdim=XSIZE(V());
+    Vdeformed().initZeros(V());
+    sumV=V().sum();
+
+    //Ip().initZeros(Xdim,Xdim);
+    Ifilteredp().initZeros(Xdim,Xdim);
+    Ifilteredp().setXmippOrigin();
+
+    // Construct mask
+    if (Rmax<0)
+    	Rmax=Xdim/2;
+    Mask mask;
+    mask.type = BINARY_CIRCULAR_MASK;
+    mask.mode = INNER_MASK;
+    mask.R1 = Rmax;
+    mask.generate_mask(Xdim,Xdim);
+    mask2D=mask.get_binary_mask();
+
+    // Low pass filter
+    filter.FilterBand=LOWPASS;
+    filter.w1=Ts/maxResol;
+    filter.raised_w=0.02;
+
+    // Transformation matrix
+    A.initIdentity(3);
+
+    createWorkFiles();
+}
+
+void ProgAngularSphAlignment::finishProcessing() {
+	XmippMetadataProgram::finishProcessing();
+	rename((fnOutDir+"/sphDone.xmd").c_str(), fn_out.c_str());
+}
+
+//#define DEBUG
+double ProgAngularSphAlignment::tranformImageSph(double *pclnm, double rot, double tilt, double psi,
+		Matrix2D<double> &A)
+{
+	const MultidimArray<double> &mV=V();
+	MultidimArray<double> &mVD=Vdeformed();
+	FOR_ALL_ELEMENTS_IN_MATRIX1D(clnm)
+		VEC_ELEM(clnm,i)=pclnm[i+1];
+	double deformation=0.0;
+	totalDeformation=0.0;
+	deformVol(mVD, mV, deformation);
+	projectVolume(mVD, P, (int)XSIZE(I()), (int)XSIZE(I()),  rot, tilt, psi);
+    double cost=0;
+	if (old_flip)
+	{
+		MAT_ELEM(A,0,0)*=-1;
+		MAT_ELEM(A,0,1)*=-1;
+		MAT_ELEM(A,0,2)*=-1;
+	}
+
+	applyGeometry(LINEAR,Ifilteredp(),Ifiltered(),A,IS_NOT_INV,DONT_WRAP,0.);
+	const MultidimArray<double> &mP=P();
+	const MultidimArray<int> &mMask2D=mask2D;
+	MultidimArray<double> &mIfilteredp=Ifilteredp();
+	double corr=correlationIndex(mIfilteredp,mP,&mMask2D);
+	cost=-corr;
+
+#ifdef DEBUG
+	std::cout << "A=" << A << std::endl;
+	Image<double> save;
+	save()=prm->P();
+	save.write("PPPtheo.xmp");
+	save()=prm->Ifilteredp();
+	save.write("PPPfilteredp.xmp");
+	save()=prm->Ifiltered();
+	save.write("PPPfiltered.xmp");
+	Vdeformed.write("PPPVdeformed.vol");
+	std::cout << "Cost=" << cost << " deformation=" << deformation << std::endl;
+	std::cout << "Press any key" << std::endl;
+	char c; std::cin >> c;
+#endif
+
+   if (showOptimization)
+   {
+		std::cout << "A=" << A << std::endl;
+		Image<double> save;
+		save()=P();
+		save.write("PPPtheo.xmp");
+		save()=Ifilteredp();
+		save.write("PPPfilteredp.xmp");
+		save()=Ifiltered();
+		save.write("PPPfiltered.xmp");
+		Vdeformed.write("PPPVdeformed.vol");
+		std::cout << "Cost=" << cost << " corr=" << corr << std::endl;
+		std::cout << "Deformation=" << totalDeformation << std::endl;
+		std::cout << "Press any key" << std::endl;
+		char c; std::cin >> c;
+    }
+
+    double massDiff=std::abs(sumV-sumVd)/sumV*100;
+    double retval=cost+lambda*(deformation+massDiff*massDiff);
+	if (showOptimization)
+		std::cout << cost << " " << deformation << " " << lambda*deformation << " " << sumV << " " << sumVd << " " << massDiff << " " << retval << std::endl;
```

Comment:
why don't you show sumV(d) in the previous block? 
Then you wouldn't need this block